### PR TITLE
New package: mnemosyne-2.11

### LINKED ISSUE #9259: Package Request: Mnemosyne
Mnemosyne a free, open-source, spaced-repetition flashcard program.  
[The build instructions are listed in the README](https://github.com/mnemosyne-proj/mnemosyne/).  

I have not managed to build it succesfully following those instructions on my musl machine; I wonder if the issue is in the setup process or there is a dependency that is broken.

### LINKED ISSUE #36997: package request: python-googletrans
https://pypi.org/project/googletrans/

### LINKED ISSUE #36998: package request: python-gTTS
https://pypi.org/project/gTTS/

### PR DESCRIPTION
**Pull request by Eloitor**
https://mnemosyne-proj.org/ flash-card tool which optimizes your learning process

#### Testing the changes
- I tested the changes in this PR: **briefly**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

#### Local build testing
- I built this PR locally for my native architecture, (x86_64)

This package builds fine. However when running mnemosyne I get this error: https://pastebin.com/mVgqnJhh

Closes #9259
Closes #36997
Closes #36998

### CODE REVIEW

**Comment by Eloitor** [comment]
I don't know how to fix this
```
=> python3-testfixtures-7.0.0_1: running do_check ...
ImportError while loading conftest '/builddir/testfixtures-7.0.0/conftest.py'.
../../conftest.py:3: in <module>
    from sybil import Sybil
E   ModuleNotFoundError: No module named 'sybil'
```
I tried adding `python3-sybil` to the depends list with no luck.

**Comment by classabbyamp** [comment]
add it to checkdepends

**Review by natrys — COMMENTED** [review]
At a minimum, the vlicenses need to happen in `post_install` block instead of `do_install`. The `do_install` block should be taken care of by the build-style, but you are overriding it here.

**Comment by Eloitor** [comment]
I'll try to update it to 2.10.1

**Comment by loreb** [comment]
This tiny patch
```
diff --git a/srcpkgs/python3-gtts/template b/srcpkgs/python3-gtts/template
index 17b4fd45..949071dc 100644
--- a/srcpkgs/python3-gtts/template
+++ b/srcpkgs/python3-gtts/template
@@ -2,8 +2,8 @@
 pkgname=python3-gtts
 version=2.3.2
 revision=1
-build_style=python3-module
-hostmakedepends="python3-setuptools"
+build_style=python3-pep517
+hostmakedepends="python3-setuptools python3-wheel"
 depends="python3 python3-click"
 checkdepends="python3-click python3-mock python3-urllib3 python3-requests python3-testfixtures python3-wheel tox"
 short_desc="Module and CLI text-to-speech tool using Google Translate API"
@@ -14,6 +14,13 @@ changelog="https://raw.githubusercontent.com/pndurette/gTTS/main/CHANGELOG.md"
 distfiles="${PYPI_SITE}/g/gTTS/gTTS-${version}.tar.gz"
 checksum=5314f7dedb4230889ff4773edd37e101408497e166982027b8d9582a4585eb4f
 
+pre_install() {
+	#dist/gTTS-2.3.2-py3-none-any.whl
+	for i in dist/gTTS-*.whl ; do
+		test -e "$i" || continue
+		cp -f "$i" "${i,,}" # lowercase
+	done
+}
 post_install() {
 	vlicense LICENSE
 }
```

and adding a couple of missing dependencies (python3-pyqt6-gui, python3-pyqt6-widgets, python3-pyqt6-printsupport, python3-pyqt6-webchannel)
leads to mnemosyne running but unable to find one of its own modules:
```
Script started on 2023-06-14 18:08:50+02:00 [TERM="rxvt-unicode-256color" TTY="/dev/pts/4" COLUMNS="80" LINES="24"]
$ mnemosyne;echo exited $?;exit
/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py:377: UserWarning: google_trans_new package is not installed. Google translate feature will not be available.
  warnings.warn("google_trans_new package is not installed. Google translate feature will not be available.")
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-_luser3'
Log body:
 An unexpected error has occurred.
Please forward the following info to the developers:

Traceback (innermost last):
  File "/bin/mnemosyne", line 312, in <module>
    mnemosyne.initialise(data_dir=data_dir, filename=filename,
  File "/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py", line 411, in initialise
    self.register_components()
  File "/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py", line 476, in register_components
    importlib.import_module(module_name), class_name)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/lib/python3.11/site-packages/mnemosyne/pyqt_ui/main_wdgt.py", line 8, in <module>
    from mnemosyne.pyqt_ui.ui_main_wdgt import Ui_MainWdgt
 ModuleNotFoundError: No module named 'mnemosyne.pyqt_ui.ui_main_wdgt'

An unexpected error has occurred.
Please forward the following info to the developers:

Traceback (innermost last):
  File "/bin/mnemosyne", line 312, in <module>
    mnemosyne.initialise(data_dir=data_dir, filename=filename,
  File "/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py", line 411, in initialise
    self.register_components()
  File "/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py", line 476, in register_components
    importlib.import_module(module_name), class_name)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/lib/python3.11/site-packages/mnemosyne/pyqt_ui/main_wdgt.py", line 8, in <module>
    from mnemosyne.pyqt_ui.ui_main_wdgt import Ui_MainWdgt
 ModuleNotFoundError: No module named 'mnemosyne.pyqt_ui.ui_main_wdgt'
An unexpected error has occurred.
Please forward the following info to the developers:

Traceback (innermost last):
  File "/bin/mnemosyne", line 312, in <module>
    mnemosyne.initialise(data_dir=data_dir, filename=filename,
  File "/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py", line 411, in initialise
    self.register_components()
  File "/usr/lib/python3.11/site-packages/mnemosyne/libmnemosyne/__init__.py", line 476, in register_components
    importlib.import_module(module_name), class_name)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/lib/python3.11/site-packages/mnemosyne/pyqt_ui/main_wdgt.py", line 8, in <module>
    from mnemosyne.pyqt_ui.ui_main_wdgt import Ui_MainWdgt
 ModuleNotFoundError: No module named 'mnemosyne.pyqt_ui.ui_main_wdgt'
exited 1

Script done on 2023-06-14 18:08:57+02:00 [COMMAND_EXIT_CODE="0"]
```

As I'm not really experienced packaging python programs, it looks to me that it's failing to import a module that is... right there???

**Comment by loreb** [comment]
Now it works, some files were supposed to be generated at build time but they weren't because... no idea why.
Anyway, starting from your latest commit:

```
diff --git a/srcpkgs/mnemosyne/template b/srcpkgs/mnemosyne/template
index a058fdae..d5d003c9 100644
--- a/srcpkgs/mnemosyne/template
+++ b/srcpkgs/mnemosyne/template
@@ -3,7 +3,7 @@ pkgname=mnemosyne
 version=2.10.1
 revision=1
 build_style=python3-module
-hostmakedepends="python3-setuptools"
+hostmakedepends="python3-setuptools python3-pyqt6-core-devel python3-pyqt6-devel-tools"
 depends="python3 python3-pyqt6 python3-pyqt6-webengine python3-Pillow python3-CherryPy
  python3-gtts python3-matplotlib python3-WebOb python3-Cheroot python3-PyOpenGL
  python3-setuptools qt6-webengine"
@@ -14,3 +14,8 @@ license="GPL-3.0-or-later"
 homepage="https://mnemosyne-proj.org/"
 distfiles="https://github.com/mnemosyne-proj/mnemosyne/archive/${version}.tar.gz"
 checksum=3691124f07335314830c943178a0b6feaf719310ad1bee582fa78726f5329276
+post_build() {
+	make build # some python files are generated
+	( cd mnemosyne/pyqt_ui && make default ) # FIXME why not already built?
+	# anyway, this is responsible for python3-pyqt6-devel-tools etc.
+}
diff --git a/srcpkgs/python3-gtts/template b/srcpkgs/python3-gtts/template
index 17b4fd45..949071dc 100644
--- a/srcpkgs/python3-gtts/template
+++ b/srcpkgs/python3-gtts/template
@@ -2,8 +2,8 @@
 pkgname=python3-gtts
 version=2.3.2
 revision=1
-build_style=python3-module
-hostmakedepends="python3-setuptools"
+build_style=python3-pep517
+hostmakedepends="python3-setuptools python3-wheel"
 depends="python3 python3-click"
 checkdepends="python3-click python3-mock python3-urllib3 python3-requests python3-testfixtures python3-wheel tox"
 short_desc="Module and CLI text-to-speech tool using Google Translate API"
@@ -14,6 +14,13 @@ changelog="https://raw.githubusercontent.com/pndurette/gTTS/main/CHANGELOG.md"
 distfiles="${PYPI_SITE}/g/gTTS/gTTS-${version}.tar.gz"
 checksum=5314f7dedb4230889ff4773edd37e101408497e166982027b8d9582a4585eb4f
 
+pre_install() {
+	#dist/gTTS-2.3.2-py3-none-any.whl
+	for i in dist/gTTS-*.whl ; do
+		test -e "$i" || continue
+		cp -f "$i" "${i,,}" # lowercase
+	done
+}
 post_install() {
 	vlicense LICENSE
 }

```

And then you'll need to install the missing deps listed above plus python3-pyqt6-multimedia, python3-pyqt6-sql and python3-argon2.
At least ItWorksForMe(tm) - let me know if it fails for you for any reason.

**Comment by loreb** [comment]
ping?

**Comment by Eloitor** [comment]
Thank you for the reminder.

I updated the packages and applied those patches, but I'm unable to build gtts...

```
./xbps-src pkg python3-gtts
....
=> python3-gtts-2.3.2_1: running do_install ...
usage: python -m installer [-h] [--destdir path] [--prefix path]
                           [--compile-bytecode level] [--no-compile-bytecode]
                           wheel
python -m installer: error: unrecognized arguments: dist/gtts-2.3.2-py3-none-any.whl
=> ERROR: python3-gtts-2.3.2_1: do_install: 'python3 -m installer --destdir ${DESTDIR} ${make_install_args} ${make_install_target}' exited with 2
=> ERROR:   in do_install() at common/build-style/python3-pep517.sh:34
```

**Comment by loreb** [comment]
It used to work... applying
```
diff --git a/srcpkgs/python3-gtts/template b/srcpkgs/python3-gtts/template
index c46e4335..a959073c 100644
--- a/srcpkgs/python3-gtts/template
+++ b/srcpkgs/python3-gtts/template
@@ -17,7 +17,7 @@ pre_install() {
 	#dist/gTTS-2.3.2-py3-none-any.whl
 	for i in dist/gTTS-*.whl ; do
 		test -e "$i" || continue
-		cp -f "$i" "${i,,}" # lowercase
+		mv -f "$i" "${i,,}" # lowercase
 	done
 }
 post_install() {
```

fixed the issue for me, but then at the end I get
```
=> mnemosyne-2.9_1: running post_build ...
make: 'build' is up to date.
./pyuic5 main_wdgt.ui > ui_main_wdgt.py
env: ‘python\r’: No such file or directory
env: use -[v]S to pass options in shebang lines
make: *** [makefile:28: ui_main_wdgt.py] Error 127
=> ERROR: mnemosyne-2.9_1: post_build: 'make default' exited with 2
=> ERROR:   in post_build() at srcpkgs/mnemosyne/template:22
=> ERROR: mnemosyne-2.9_1: post_build: '( cd mnemosyne/pyqt_ui && make default )' exited with 1
=> ERROR:   in post_build() at srcpkgs/mnemosyne/template:22
```

Apparently masterdir/builddir/mnemosyne-2.9/mnemosyne/pyqt_ui/pyuic5 has a carriage return in its shabang (fixed upstream, but there's also a new release, it's now pyuic6)

Fix that and it now complains (it used to work I swear!!!) that
```
./pyuic5 add_cards_dlg.ui > ui_add_cards_dlg.py
Traceback (most recent call last):
  File "/builddir/mnemosyne-2.9/mnemosyne/pyqt_ui/./pyuic5", line 15, in <module>
    from PyQt5 import QtCore
ModuleNotFoundError: No module named 'PyQt5'
```

The newer release uses PyQt6, so anything we do to fix that (nothing difficult, every time I try to build a new tiny patch/dependency is needed) will have to be discarded when we upgrade - it says above you tried 2.10, did anything go wrong? I still have hostdir/binpkgs/mnemosyne/mnemosyne-2.10.1_1.x86_64-musl.xbps, why downgrading to 2.9?

File: srcpkgs/mnemosyne/template
Comment (loreb) [inline srcpkgs/mnemosyne/template:1]: There's a duplicate "depends="

**Comment by loreb** [comment]
Just to be sure: it compiles & runs but the tests fail because of missing dependencies, is that correct?

**Comment by Eloitor** [comment]
The version downgrade is bc I got confused in one of the multiple attempts. I now attempted using the makefile instead of setup.py. Now everything is built. Can you check it out?

**Comment by loreb** [comment]
It compiles and installs, but when you run it complains about missing imports; I trusted my comment above, manually installed python3-pyqt6-multimedia, python3-pyqt6-sql and python3-argon2, and it works (ie I tried a couple of questions)

**Comment by Eloitor** [comment]
I removed the package and installed again... Then it worked

**Comment by Eloitor** [comment]
I had to manually install the dependencies: `python3-pyqt6-network python3-pyqt6-gui python3-pyqt6-widgets python3-pyqt6-printsupport python3-pyqt6-webchannel python3-pyqt6-multimedia python3-pyqt6-sql`. I'll add them later to the template.

**Comment by Eloitor** [comment]
qt6-plugin-sqlite is also necessary for the "browse cards" button to work

**Comment by Eloitor** [comment]
Now it should work without installing anything else manually. :) Thank you!
Now let's try to fix the problems with tests.

**Comment by Eloitor** [comment]
Now tests work locally except for python3-testfixtures. I don't know how to fix it.

I tried to disable CI checks for python3-sybil and python3-testfixtures, but it keeps complaining about the cyclic dependency. I don't know how to do it properly.

**Comment by loreb** [comment]
I can think of a few disgusting hacks (say, making two pairs of identical packages, such that sybil is tested with fixtures-dummy, fixtures is tested with sybil-dummy, and the -dummy packages have their do_check be a NOP), but no clean solution either.

Can you try asking on IRC/reddit?

**Comment by Eloitor** [comment]
I removed checks for python3-testfixtures. If that's ok let me know and I'll squash the commits for python3-testfixtures together.

**Comment by Eloitor** [comment]
I just realized that sybil was only required to be able to test python3-testfixtures. If I disable the check for testfixtures it doesn't make sense to package it.... I'll attempt to remove the check for sybil and enable it for testfixtures

**Comment by Eloitor** [comment]
I don't know how to fix this check error on python3-testfixtures

```
________ ERROR collecting build/lib/testfixtures/tests/test_compare.py _________
import file mismatch:
imported module 'testfixtures.tests.test_compare' has this __file__ attribute:
  /builddir/python3-testfixtures-7.1.0/testfixtures/tests/test_compare.py
which is not the same as the test file we want to collect:
  /builddir/python3-testfixtures-7.1.0/build/lib/testfixtures/tests/test_compare.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

**Comment by loreb** [comment]
The following (intentionally dirty) diff did it for me:
```
diff --git a/srcpkgs/python3-testfixtures/template b/srcpkgs/python3-testfixtures/template
index 6c71aad1..121bad83 100644
--- a/srcpkgs/python3-testfixtures/template
+++ b/srcpkgs/python3-testfixtures/template
@@ -12,10 +12,11 @@ license="MIT"
 homepage="https://github.com/Simplistix/testfixtures"
 distfiles="${PYPI_SITE}/t/testfixtures/testfixtures-${version}.tar.gz"
 checksum=456ce4f393160f27cd6840a5370ecf4a70f1721737f5e67a2af79e6c817804a4
+export PYTHONDONTWRITEBYTECODE=1 # NOPE
 
 do_check() {
-	cd build/lib
-	pytest -v --ignore="testfixtures/tests/test_django" # Fails to import
+	cd testfixtures &&
+	pytest -v --ignore="tests/test_django" # Fails to import
 }
 
 post_install() {
```

RANT: based on stackoverflow, I tried the python bytecode thing (did nothing, I just left it as a reminder that it's useless),
then I thought I understood when I read https://docs.pytest.org/en/latest/explanation/goodpractices.html#test-discovery but nope, no way...
I still don't know why it went all the way up instead of staying in build/lib, I expected it to stop at lib (you know, the first directory without __init__.py), but if it works it works.

And if you need more circular fun... at first I thought maybe they fixed the bug upstream, I tried pytest 7.4.2 and boom! circular dependencies.

Sorry for the rant, I just needed it.

**Comment by Eloitor** [comment]
Amazing! thank you

**Comment by loreb** [comment]
bump!

**Comment by Eloitor** [comment]
I updated it to 2.11, it works on my machine now

**Comment by loreb** [comment]
bump!

Why is this failing to build all of a sudden?

**Comment by Eloitor** [comment]
I do not have enough time to try to fix it... If you want feel free to try to fix it and create a new PR

**Pull request closed without merging**